openedx-unsupported/ecommerce
# ENT-1281 Add enterprise coupon backend api

**Description**
This PR introduces new api endpoints for creating and updating Enterprise Coupons, which inherit from the existing Coupon endpoints. We also ensure that based on the enterprise coupon switch, that the original coupon endpoints will not accept or return enterprise data. In order to aid with the transition, we have disabled the patch option for the coupon screen so that all fields get sent for updates instead of just changed fields.

**Test Instructions**
Deployed on the business sandbox.

_With the ENTERPRISE_OFFERS_FOR_COUPONS_SWITCH off:_
- Creating and updating coupons without enterprise data should be unaffected.
- Creating a coupon with enterprise data will result in two conditional offers created per voucher, with the enterprise conditional offer having a condition with proxy_class of AssignableEnterpriseCustomerCondition. The client name will be the Enterprise name even if a client name is supplied. The Business Client model will have the enterprise uuid field populated.
- Updating a coupon with enterprise data should result in an additional enterprise conditional offer created or updated, and the client name will be the enterprise name. Even updates with no data changes should trigger these updates. This will allow us to migrate coupons no matter what enterprise data has already been set.

_With the ENTERPRISE_OFFERS_FOR_COUPONS_SWITCH on:_
- Creating and updating coupons without enterprise data should be unaffected.
- All "migrated" enterprise coupons will not show in the list view. A "migrated" enterprise coupon is one that has the attached business client model with the enterprise uuid set, as well as enterprise conditional offers for each voucher available.
- Non "migrated" enterprise coupons can still be updated, but once updated, they will not show up in the list view.
- New coupons with enterprise data cannot be created from the coupons screen.

_Note:_ The backend apis for CRUD operations on enterprise coupons are created in this PR, but the new coupon screen UI that utilizes those apis is not included in this change set, so test manual testing instructions for exercising those endpoints will be included in the PR that has the coupon screen UI.
